tour_model_eval/v-measurel_cutoff_bins_single_user.ipynb

```diff
@@ -0,0 +1,271 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "thorough-nomination",
+   "metadata": {},
+   "source": [
+    "## Choose one user for experiment"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "revolutionary-sussex",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "user = all_users[4]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "crude-compatibility",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#read the data from the database. We choose key='confirmed_trip' to get confirmed trips here\n",
+    "trips = pipeline.read_data(uuid=user,key=esda.CONFIRMED_TRIP_KEY)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "right-graham",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# select trips that have user_input to analyze\n",
+    "non_empty_trips = [t for t in trips if t[\"data\"][\"user_input\"] != {}]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "infrared-accordance",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "len(non_empty_trips), non_empty_trips"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "generic-glasgow",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "bin_trips, bins = pipeline.remove_noise(non_empty_trips, radius)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "thick-commissioner",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "logging.debug('The list of bins is %s' % bins)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "macro-round",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# show all user labels in all bins\n",
+    "for bin in bins:\n",
+    "    bin_user_input = (non_empty_trips[i].data[\"user_input\"] for i in bin)\n",
+    "    bin_df = pd.DataFrame(data = bin_user_input)\n",
+    "    print(bin_df)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "tired-athens",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# turn all user_input into list without binning\n",
+    "bin_trips_user_input_ls = pd.DataFrame(data=[bin_trips[i][\"data\"][\"user_input\"] for i in range(len(bin_trips))]).values.tolist()\n",
+    "bin_trips_user_input_ls"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "veterinary-courage",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for i in bin_trips_user_input_ls:\n",
+    "    if i[2] == \"same_mode\":\n",
+    "        print(i)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "coastal-fiction",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# drop duplicate user_input\n",
+    "no_dup_df=pd.DataFrame(data=[bin_trips[i][\"data\"][\"user_input\"] for i in range(len(bin_trips))]).drop_duplicates()\n",
+    "no_dup_df"
+   ]
```

Comment:
Then this is just:
```suggestion
   "source": [
    "# drop duplicate user_input\n",
    "no_dup_df= bin_trips_df.drop_duplicates()\n",
    "no_dup_df"
   ]
```

Switching back and forth between pandas dataframes and lists is inefficient - the dataframes are designed to be efficient for multiple sequential operations.